flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java

```diff
@@ -0,0 +1,573 @@
+    public void testDeriveEntryClassInformationFromSystemClasspathWithNonExistingJobClassName()
+            throws IOException, FlinkException {
+        classpathProvider.setSystemClasspathWithEntryClass();
+
+        final String jobClassName = "SomeJobClassNotBeingOnTheSystemClasspath";
+        final EntryClassInformationProvider informationProvider =
+                PackagedProgramRetrieverImpl.createEntryClassInformationProvider(
+                        null, null, null, jobClassName, new String[0]);
+        assertThat(informationProvider.getJobClassName().isPresent(), is(true));
+        assertThat(informationProvider.getJobClassName().get(), is(jobClassName));
+        assertThat(informationProvider.getJarFile().isPresent(), is(false));
+    }
+
+    @Test
+    public void testDeriveEntryClassInformationFromSystemClasspathWithExistingJobClassName()
+            throws IOException, FlinkException {
+        classpathProvider.setSystemClasspathWithEntryClass();
+
+        final EntryClassInformationProvider informationProvider =
+                PackagedProgramRetrieverImpl.createEntryClassInformationProvider(
+                        null, null, null, classpathProvider.getJobClassName(), new String[0]);
+        assertThat(informationProvider.getJobClassName().isPresent(), is(true));
+        assertThat(
+                informationProvider.getJobClassName().get(),
+                is(classpathProvider.getJobClassName()));
+        assertThat(informationProvider.getJarFile().isPresent(), is(false));
+    }
+
+    @Test
+    public void testDeriveEntryClassInformationFromSystemClasspathExtractingTheJobClassFromThere()
+            throws IOException, FlinkException {
+        classpathProvider.setSystemClasspathWithEntryClass();
+
+        final EntryClassInformationProvider informationProvider =
+                PackagedProgramRetrieverImpl.createEntryClassInformationProvider(
+                        null, null, null, null, new String[0]);
+        assertThat(informationProvider.getJobClassName().isPresent(), is(true));
+        assertThat(
+                informationProvider.getJobClassName().get(),
+                is(classpathProvider.getJobClassName()));
+        assertThat(informationProvider.getJarFile().isPresent(), is(false));
+    }
+
+    @Test
+    public void testDeriveEntryClassInformationFromClasspathWithJobClass()
+            throws IOException, FlinkException {
+        final EntryClassInformationProvider informationProvider =
+                PackagedProgramRetrieverImpl.createEntryClassInformationProvider(
+                        // the user directory must be specified
+                        classpathProvider.getUserDirectoryWithTwoEntryClasses(),
+                        // the user classpath is derived from the user directory outside of the
+                        // method
+                        classpathProvider.getURLUserClasspathWithTwoEntryClasses(),
```

Comment:
Aren't these irrelevant since we explicitly provide the job classname anyway?

Reply:
the comments are also weird; they claim the user directory is required, but testDeriveEntryClassInformationFromSystemClasspathWithNonExistingJobClassName uses null for it and does the same assertions.

Reply:
There's a slight difference between providing a user classpath through the user directory and not providing one. In the former case, the code checks whether the job class appears in the user classpath. For the latter one, this check is not done.

That's the difference between `testDeriveEntryClassInformationFromClasspathWithJobClass` and `testDeriveEntryClassInformationFromSystemClasspathWithNonExistingJobClassName`.